File: Dynamic Programming/91_decode_ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = [0 for _ in range(len(s) + 1)]
        dp[0] = 1
        dp[1] = 1 if s[0] != "0" else 0

        for i in range(2, len(dp)):
            if s[i - 1] != "0": 
                dp[i] = dp[i - 1]
            
            two_digit = int(s[i - 2: i])
            if two_digit >= 10 and two_digit <= 26:
                dp[i] += dp[i - 2]

        return dp[-1]

File: Dynamic Programming/91_decode_ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        if(s[0] == '0') return 0;
        
        int one = 1, two = 1;

        for(int i = 2; i <= s.size(); i += 1) {
            int temp = two;
            if(s[i - 1] == '0') two = 0;
            
            int two_digit = stoi(s.substr(i - 2, 2));
            if(two_digit >= 10 && two_digit <= 26) two += one;
            
            one = temp;
        }

        return two;
    }
};